### PR TITLE
[AV-231] Runge-Kutta Engine

### PR DESCRIPTION
**Pull request by JacobGugala**
New physics engine that is more accurate than Forward Euler.  Some things to keep in mind:
- the 2 helper methods use the same calculations as in Forward Euler (for comparison)
- would it make more sense to rename "{variable}_avg" to "{variable}_rk4"?
- k4 math is the most uncertain
- links to the math are on trello tickets

### CODE REVIEW

**Review by NPIPHI — CHANGES_REQUESTED** [review]
This is a big step for the physics engine! Mostly reviewed good practices/style stuff, I'm going to review the math once I've read up on it.

File: includes/PhysicsEngine.h
Comment (AyberkY) [inline includes/PhysicsEngine.h:54]: The name `Struct` is quite generic, and I can foresee it conflicting with other software components in the future.

Since this struct is specific to your RK4 algorithm, you should define this struct in the `private:` sector of the `RungeKutta` class. That way the name `State` is only relevant within the class implementation.

While you're at it, I'd also suggest renaming this to `RungeKuttaState` for the sake of verbosity, but that's up to you

File: src/PhysicsEngine.cpp
Comment (AyberkY) [inline src/PhysicsEngine.cpp:193]: This docstring needs some work.

I think the `@brief` is not accurate. I'd suggest: "Performs a fourth-order Runge Kutta method to advance the physics forward one timestep."

After `@brief` you can add more information in the form of paragraphs. Since this is a math heavy function, you should put some sources of where you got the math from. Check out the function I wrote in [[AV-346] PR](https://github.com/ISSUIUC/ISS_SILSIM/pull/19) as an example

Also please briefly explain what the `@param`s are in one line

File: src/PhysicsEngine.cpp
Comment (AyberkY) [inline src/PhysicsEngine.cpp:270]: We can see what the parameters and returns are ourselves lol. Please add a short description of what these are

File: src/PhysicsEngine.cpp
Comment (AyberkY) [inline src/PhysicsEngine.cpp:326]: typo

File: src/PhysicsEngine.cpp
Comment (AyberkY) [inline src/PhysicsEngine.cpp:330]: As before, please explain

File: src/PhysicsEngine.cpp
Comment (AyberkY) [inline src/PhysicsEngine.cpp:440]: Please copy my implementation of this exactly from [[AV-346]](https://github.com/ISSUIUC/ISS_SILSIM/pull/19) for consistency including the docstring.

**Review by AyberkY — CHANGES_REQUESTED** [review]
Lookin very good. Nice work.

I have a couple small tweaks, but otherwise the implementation checks out.

Also keep in mind that [AV-294](https://github.com/ISSUIUC/ISS_SILSIM/pull/21) is migrating to use the Eigen library. If that PR merges before this, you'll have merge conflicts and you'll need to switch to using Eigen as well before merging.

Also don't forget to run `clang-format`

**Comment by AyberkY** [comment]
Oh yeah, @JacobGugala don't forget to add yourself as an author to `PhysicsEngine.h/.cpp`

File: src/PhysicsEngine.cpp
Comment (AyberkY) [reply src/PhysicsEngine.cpp]: Since when? My compiler definitely disagrees 🤔

File: src/PhysicsEngine.cpp
Comment (NPIPHI) [inline src/PhysicsEngine.cpp:214]: k1 needs to be recalculated with the calc_state function here. vel_if and friends are calculated before all the derivatives from the previous iteration are applied so the derivative needs to be recomputed.

**Review by NPIPHI — CHANGES_REQUESTED** [review]

**Comment by AyberkY** [comment]
FYI I just merged [[AV-346]](https://github.com/ISSUIUC/ISS_SILSIM/pull/19) PR.

File: src/PhysicsEngine.cpp
Comment (JacobGugala) [reply src/PhysicsEngine.cpp]: You can't call calc_state for k1.  One of the parameters for calc_state is a State, since the function bases its calculations on the initial values pulled from the rocket class and the information stored in the State class.  Since k1 is the state the rocket is starting in, this function can't be called because no other States exist to pass as a parameter.  It also does not need to be recalculated since k1 is just the state that the rocket is currently in, which was calculated by the last RungeKutta step and doing that math again would be redundant.

**Review by NPIPHI — APPROVED** [review]

File: src/PhysicsEngine.cpp
Comment (AyberkY) [inline src/PhysicsEngine.cpp:428]: I'm still not a fan of writing intermediate values into the Rocket class and then resetting it back to default. Feels like bad software design IMO.

I see that we only have to do this because the `calc_net_*` functions use the `r2i/i2r` functions of the rocket class. I think in the future we should have `r2i/i2r` helper functions in `PhysicsEngine` as well that take take in whatever quaternion you want to pass in.

I don't think this blocks merging the PR, but if you agree, then please make a ticket to implement this later.

File: src/PhysicsEngine.cpp
Comment (AyberkY) [inline src/PhysicsEngine.cpp:290]: I think `calc_net_force` and `calc_net_torque` should be members of the `PhysicsEngine` base class and shared among the Euler and RK4 engines.

I don't think this blocks merging this PR, but please make a ticket for this if you agree.

File: src/PhysicsEngine.cpp
Comment (AyberkY) [inline src/PhysicsEngine.cpp:242]: clang-format really doesn't like this line in particular lol.

**Review by AyberkY — APPROVED** [review]
I have a few comments that suggest small design improvements we could make in the future. But otherwise I approve.

@NPIPHI you should add yourself as an author too.

**Pull request merged**
